// appium-ios-simulator/build/lib/tail-until.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _teen_process = require('teen_process');

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

// tails a file, promise resolves when input string is written to file
function tailUntil(filePath, until) {
  var timeout = arguments.length <= 2 || arguments[2] === undefined ? 5000 : arguments[2];
  var proc, startDetector;
  return _regeneratorRuntime.async(function tailUntil$(context$1$0) {
    var _this = this;

    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        proc = new _teen_process.SubProcess('tail', ['-f', '-n', '100', filePath]);

        startDetector = function startDetector(stdout) {
          return stdout.indexOf(until) > -1;
        };

        return context$1$0.abrupt('return', new _bluebird2['default'](function callee$1$0(resolve, reject) {
          var started, timedout;
          return _regeneratorRuntime.async(function callee$1$0$(context$2$0) {
            while (1) switch (context$2$0.prev = context$2$0.next) {
              case 0:
                started = proc.start(startDetector);
                timedout = _bluebird2['default'].delay(timeout).then(function () {
                  return reject('tailing file ' + filePath + ' failed after ' + timeout + 'ms');
                });
                context$2$0.next = 4;
                return _regeneratorRuntime.awrap(_bluebird2['default'].race([started, timedout]));

              case 4:

                resolve();

              case 5:
              case 'end':
                return context$2$0.stop();
            }
          }, null, _this);
        })['finally'](function callee$1$0() {
          return _regeneratorRuntime.async(function callee$1$0$(context$2$0) {
            while (1) switch (context$2$0.prev = context$2$0.next) {
              case 0:
                if (!proc.isRunning) {
                  context$2$0.next = 9;
                  break;
                }

                context$2$0.prev = 1;
                context$2$0.next = 4;
                return _regeneratorRuntime.awrap(proc.stop());

              case 4:
                context$2$0.next = 9;
                break;

              case 6:
                context$2$0.prev = 6;
                context$2$0.t0 = context$2$0['catch'](1);

                // there is not much we can do here, unfortunately, but log
                _logger2['default'].info('Stopping tail process failed: ' + context$2$0.t0.message);

              case 9:
              case 'end':
                return context$2$0.stop();
            }
          }, null, _this, [[1, 6]]);
        }));

      case 3:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

exports.tailUntil = tailUntil;

// // for debugging
// function consoleOut (...args) {
//   console.log(`>>> ${args}`); // eslint-disable-line no-console
// }
// proc.on('output', consoleOut);

/* eslint-disable promise/prefer-await-to-then */

/* eslint-enable */

// no matter what, stop the tail process
